p2p/security/noise/transport.go

```diff
@@ -65,3 +82,50 @@ func (t *Transport) WithSessionOptions(opts ...SessionOption) (sec.SecureTranspo
 	}
 	return st, nil
 }
+
+func MatchMuxers(initiatorMuxers, responderMuxers []string) string {
+	respMuxerMap := make(map[string]struct{}, len(responderMuxers))
+	for _, muxer := range responderMuxers {
+		respMuxerMap[muxer] = struct{}{}
+	}
+
+	selectedMuxer := ""
+	for _, muxer := range initiatorMuxers {
+		if _, ok := respMuxerMap[muxer]; ok {
+			selectedMuxer = muxer
+			break
+		}
+	}
+	return selectedMuxer
+}
+
+type TransportEarlyDataHandler struct {
+	transport      *Transport
+	receivedMuxers []string
+}
+
+func NewTransportEDH(t *Transport) *TransportEarlyDataHandler {
+	return &TransportEarlyDataHandler{transport: t}
+}
+
+func (i *TransportEarlyDataHandler) Send(context.Context, net.Conn, peer.ID) *pb.NoiseExtensions {
+	return &pb.NoiseExtensions{
+		WebtransportCerthashes: [][]byte{},
+		StreamMuxers:           i.transport.muxers,
+	}
+}
+
+func (i *TransportEarlyDataHandler) Received(_ context.Context, _ net.Conn, extension *pb.NoiseExtensions) error {
+	// Discard messages with size or the number of protocols exceeding extension limit for security.
+	if extension != nil && extension.XXX_Size() <= MAX_EXTENSION_SIZE && len(extension.StreamMuxers) <= MAX_PROTO_NUM {
```

Comment:
We need to check the extension size _before_ we parse it.

Reply:
At the time this function is called, the protobuf is already parsed (otherwise we wouldn't have a `pb.NoiseExtensions`).

Reply:
I see, you mean check the size before the whole handshake message is parsed, right? 
That is more general a case that relates to the whole handshake process.  We can either piggy-back the handshaek message size check in this PR or add it as a separate PR for easier tracking. I am OK with either option.

Reply:
as per discussion on size check with uint16, this is resolved.